File: src/routes/customer.routes.ts
import { Hono } from "hono";
import { Env } from "@/types/hono";
import { CustomerService } from "@/services/customer.service";
import { successResponse, paginatedResponse } from "@/utils/response";
import { listCustomersQuerySchema } from "@/types/customer/customer";
import { authMiddleware } from "@/middleware/auth";
import { toResponse, toResponseArray } from "@/utils/response-transformer";

export const customerRoutes = new Hono<Env>();

// Lazy initialization of service
let customerService: CustomerService | null = null;

function getCustomerService(): CustomerService {
  if (!customerService) {
    customerService = new CustomerService();
  }
  return customerService;
}

// Apply auth middleware to all customer routes
customerRoutes.use("*", authMiddleware);

// GET /api/customers - List all customers
customerRoutes.get("/", async (c) => {
  const user = c.get("user");
  const query = listCustomersQuerySchema.parse(
    Object.fromEntries(new URL(c.req.url).searchParams)
  );

  const { customers, total } = await getCustomerService().listCustomers(
    user.companyId!,
    {
      page: query.page,
      limit: query.limit,
      status: query.status,
      search: query.search,
    }
  );

  return c.json(paginatedResponse(toResponseArray(customers), query.page, query.limit, total));
});

// GET /api/customers/all - List all customers (without pagination)
customerRoutes.get("/all", async (c) => {
  const user = c.get("user");
  const customers = await getCustomerService().getAllCustomers(user.companyId!);
  return c.json(successResponse(toResponseArray(customers)));
});

// GET /api/customers/:id - Get single customer
customerRoutes.get("/:id", async (c) => {
  const user = c.get("user");
  const id = c.req.param("id");

  const customer = await getCustomerService().getCustomer(user.companyId!, id);

  return c.json(successResponse(toResponse(customer)));
});

// POST /api/customers - Create new customer
customerRoutes.post("/", async (c) => {
  const user = c.get("user");
  const body = await c.req.json();

  const customer = await getCustomerService().createCustomer(user.companyId!, body);

  return c.json(successResponse(toResponse(customer)), 201);
});

// PUT /api/customers/:id - Update customer
customerRoutes.put("/:id", async (c) => {
  const user = c.get("user");
  const id = c.req.param("id");
  const body = await c.req.json();

  const customer = await getCustomerService().updateCustomer(user.companyId!, id, body);

  return c.json(successResponse(toResponse(customer)));
});

// DELETE /api/customers/:id - Delete customer
customerRoutes.delete("/:id", async (c) => {
  const user = c.get("user");
  const id = c.req.param("id");

  await getCustomerService().deleteCustomer(user.companyId!, id, user.role);

  return c.json(successResponse({ message: "Customer deleted successfully" }));
});

File: fisoft/src/main/kotlin/routes/CustomerRoutes.kt
package com.ilyasdemirkiran.routes

import com.ilyasdemirkiran.middleware.authenticateUser
import com.ilyasdemirkiran.middleware.currentCompanyId
import com.ilyasdemirkiran.middleware.currentUser
import com.ilyasdemirkiran.services.CustomerImageService
import com.ilyasdemirkiran.services.CustomerNoteService
import com.ilyasdemirkiran.services.CustomerService
import com.ilyasdemirkiran.types.UserRole
import com.ilyasdemirkiran.utils.paginatedResponse
import com.ilyasdemirkiran.utils.successResponse
import io.ktor.http.*
import io.ktor.http.content.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class CreateCustomerReq(
    val name: String,
    val surname: String,
    val phone: String? = null,
    val city: String? = null,
    val district: String? = null,
    val address: String? = null
)

@Serializable
data class CreateNoteReq(val note: String)

fun Route.customerRoutes() {
    val customerService = CustomerService()
    val noteService = CustomerNoteService()
    val imageService = CustomerImageService()

    route("/api/customers") {
        get {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1
            val limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20
            val status = call.request.queryParameters["status"]
            val search = call.request.queryParameters["search"]

            val (customers, total) = customerService.listCustomers(companyId, page, limit, status, search)
            call.respond(HttpStatusCode.OK, paginatedResponse(customers, page, limit, total))
        }

        get("/all") {
            call.authenticateUser()
            val user = call.currentUser
            val companyId = call.currentCompanyId
            val status = if (user.role == UserRole.user) "active" else null
            val customers = customerService.getAllCustomers(companyId, status)
            call.respond(HttpStatusCode.OK, successResponse(customers))
        }

        post {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val req = call.receive<CreateCustomerReq>()
            val customer = customerService.createCustomer(companyId, req.name, req.surname, req.phone, req.city, req.district, req.address)
            call.respond(HttpStatusCode.Created, successResponse(customer))
        }

        get("/{id}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val id = call.parameters["id"] ?: ""
            val customer = customerService.getCustomer(companyId, id)
            call.respond(HttpStatusCode.OK, successResponse(customer))
        }

        put("/{id}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val id = call.parameters["id"] ?: ""
            val req = call.receive<CreateCustomerReq>()
            val updated = customerService.updateCustomer(companyId, id, req.name, req.surname, req.phone, req.city, req.district, req.address)
            call.respond(HttpStatusCode.OK, successResponse(updated))
        }

        delete("/{id}") {
            call.authenticateUser()
            val user = call.currentUser
            val companyId = call.currentCompanyId
            val id = call.parameters["id"] ?: ""
            customerService.deleteCustomer(companyId, id, user.role)
            call.respond(HttpStatusCode.OK, successResponse(mapOf("success" to true)))
        }

        // Customer Notes
        get("/{id}/notes") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val id = call.parameters["id"] ?: ""
            val notes = noteService.getCustomerNotes(companyId, id)
            call.respond(HttpStatusCode.OK, successResponse(notes))
        }

        post("/{id}/notes") {
            call.authenticateUser()
            val user = call.currentUser
            val companyId = call.currentCompanyId
            val id = call.parameters["id"] ?: ""
            val req = call.receive<CreateNoteReq>()
            val note = noteService.createNote(companyId, id, user._id, req.note)
            call.respond(HttpStatusCode.Created, successResponse(note))
        }

        put("/{id}/notes/{noteId}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val noteId = call.parameters["noteId"] ?: ""
            val req = call.receive<CreateNoteReq>()
            val updated = noteService.updateNote(companyId, noteId, req.note)
            call.respond(HttpStatusCode.OK, successResponse(updated))
        }

        delete("/{id}/notes/{noteId}") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val noteId = call.parameters["noteId"] ?: ""
            noteService.deleteNote(companyId, noteId)
            call.respond(HttpStatusCode.OK, successResponse(mapOf("message" to "Customer note deleted successfully")))
        }

        // Customer Images by Customer ID
        get("/{id}/images") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val id = call.parameters["id"] ?: ""
            val images = imageService.getCustomerImages(companyId, id)
            call.respond(HttpStatusCode.OK, successResponse(images))
        }

        post("/{id}/images") {
            call.authenticateUser()
            val companyId = call.currentCompanyId
            val customerId = call.parameters["id"] ?: ""
            val multipart = call.receiveMultipart()
            var filename = "image.jpg"
            var mimeType = "image/jpeg"
            var bytes: ByteArray? = null
            var labels: List<String> = emptyList()

            multipart.forEachPart { part ->
                when (part) {
                    is io.ktor.http.content.PartData.FileItem -> {
                        filename = part.originalFileName ?: "image.jpg"
                        mimeType = part.contentType?.toString() ?: "image/jpeg"
                        bytes = part.streamProvider().readBytes()
                    }
                    is io.ktor.http.content.PartData.FormItem -> {
                        if (part.name == "labels") {
                            labels = part.value.split(",").map { it.trim() }.filter { it.isNotEmpty() }
                        }
                    }
                    else -> {}
                }
                part.dispose()
            }

            if (bytes == null) {
                call.respond(HttpStatusCode.BadRequest, mapOf("success" to false, "error" to "No file provided"))
                return@post
            }

            val image = imageService.uploadImage(companyId, customerId, filename, bytes!!, mimeType, labels)
            call.respond(HttpStatusCode.Created, successResponse(image))
        }
    }
}
